### DISCUSSION TITLE
Idea: Ghostty for iPad

### DISCUSSION THREAD

**Discussion by cschaba** [ghostty-org/ghostty#4087] [Feature Requests, Ideas] [answered]
Currently I am using "Blink" and "Shellfish" on my iPad.  I would really like to see Ghostty on my IPadPro as well as on my Macbook.  

Maybe someone can find the time to make a port.

**Comment by mitchellh** [comment, marked as answer]
`libghostty` will enable others to do this, I don't have plans currently to do this.

**Comment by kitknox** [comment]
So I went and did the thing.  Much thanks to @mitchellh for all of the ground work that was laid.  I have written a new Swift/SwiftUI/UIKit front end and have it wrapping `libghostty` and running on iOS, iPadOS, visionOS, and macOS.  While libghostty was already in a state that it compiled on iOS, it did not actually work on iOS.  The kevent queues in `libxev` didn't work on iOS without fixes.  This wasn't an obvious fix as initially these were all async crashes and then because my first attempt at resolving this ended up resulting in fixing the crashes but no glyphs rendering.  I was just left with a cursor.  It didn't help that I hadn't touched a line of zig previously.  `libghostty` itself I added a new pipes based backend type as we can't exec() and don't have PTYs in our sandboxed iOS environment.  For getting the visionOS builds up and running I also ran into trouble with the zig stdlib itself and had to patch that.  So I now have a backlog of things I need to sort through cleaning up and figuring out upstream PRs for.  I think this is the first time there is a native terminal app for visionOS that isn't just something running in iPad compatibility mode?

SSH sits on top of Swift-NIO-SSH.  Scrolling, splits, copy/paste, resizing, moving tabs and the major foundational pieces I have in place so far.  There is also a basic local shell that is compatible with the iOS sandbox.

@mitchellh I was hoping to get input on the concept of a pipe based backend and if that makes sense to you as the correct approach. It seems to work and perform quite well so far.  It would be great to not have to maintain a separate fork on an ongoing basis.

Stay tuned as I get project pages out, an initial Test Flight etc.  Preview screenshots :

![9E2C33C5-D891-48B2-A1BE-3F33DC714B18_1_102_o](https://github.com/user-attachments/assets/b75c86e7-337e-401e-af24-34b3d185acc6)
![7B78DAE6-8343-4490-BE1A-3CBB79697E9A_1_105_c](https://github.com/user-attachments/assets/faa2d4b0-57b4-4b96-b732-ff14e8f93231)
![27745EBB-5AA9-4FFF-BC8A-9614885F6DF8_1_105_c](https://github.com/user-attachments/assets/1d50d2ef-b7f3-421f-b47f-76da4021fca9)
![33CF050C-1160-4F59-B042-D9169564CAAC_1_101_o](https://github.com/user-attachments/assets/c22ad3fd-f4e5-46bf-8f67-db6b88e89fb3)

**Reply by mitchellh** [reply]
Amazing work! Yes, a pipe-based backend would be accepted (assuming the design is good). That was always the thought with the `termio` package, to provide alternate input mechanisms or abstract it more in a way so that exec vs pipe was possible, especially so we can support remote instances.

**Reply by realhackcraft** [reply]
This is really cool! How’s your progress on this @kitknox ?

**Reply by kitknox** [reply]
TestFlight links are live at [beta.rootshell.com](https://beta.rootshell.com)

One of the things I was going to circle back to here is I also ended up needing to implement CADisplayLink in order to get rough parity in terms of GPU use with the regular macOS builds of ghostty since CVDisplayLink is both deprecated now and wasn't available on all of the new build targets.  It seems it was falling back to just a timer for rendering and doing a lot of extra work.  I don't know if there were already plans underway given the deprecation.

**Comment by aspauldingcode** [comment]
@kitknox  I ported zsh, fastfetch, coreutils to iOS with AppStore compliance. 
Would you like to implement it?
